### PR TITLE
feat(pubsub): Process message events and forward matching triggers.

### PR DESCRIPTION
**Pull request by jtk54**
Adds specs for a crucial classes in this code path. Also refactors the 10+ arg `Trigger` constructors into `builder`s.

### CODE REVIEW

File: echo-model/src/main/java/com/netflix/spinnaker/echo/model/Trigger.java [hunk 1]
```diff
@@ -69,29 +71,80 @@ public String toString() {
   String branch;
   String runAsUser;
   String secret;
+  String subscriptionName;
+  String pubsubType;
 
   public Trigger atBuildNumber(final int buildNumber) {
-    return new Trigger(enabled, id, type, master, job, buildNumber, propertyFile, cronExpression, source, project, slug, null, account, repository, null, digest, null, branch, runAsUser, secret);
+    return this.toBuilder()
```
Comment (lwander) [inline echo-model/src/main/java/com/netflix/spinnaker/echo/model/Trigger.java]: Noice

File: echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/PubsubEventMonitor.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+  @Override
+  protected Predicate<Trigger> matchTriggerFor(final TriggerEvent event) {
+    PubsubEvent pubsubEvent = (PubsubEvent) event;
+    MessageDescription description = pubsubEvent.getContent().getMessageDescription();
+
+    // TODO(jacobkiefer): Need to apply filters specified in triggers here.
+    return trigger -> trigger.getType().equalsIgnoreCase(PUBSUB_TRIGGER_TYPE)
+        && trigger.getPubsubType().equalsIgnoreCase(description.getPubsubType().toString())
+        && trigger.getSubscriptionName().equalsIgnoreCase(description.getSubscriptionName());
```
Comment (lwander) [inline echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/PubsubEventMonitor.java]: If you have multiple files in one GCS bucket, do each have their own subscription?

File: echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/PubsubEventMonitor.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  }
+
+  @Override
+  protected void onMatchingPipeline(Pipeline pipeline) {
+    super.onMatchingPipeline(pipeline);
+    val id = registry.createId("pipelines.triggered")
```
Comment (lwander) [inline echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/PubsubEventMonitor.java]: Why is this happening outside the `isPubsubTrigger` guard below? Will this be applied to all triggered pipelines, not just pubsub ones?

File: echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/PubsubEventMonitor.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+      id.withTag("pubsubType", pipeline.getTrigger().getPubsubType());
+      id.withTag("subscriptionName", pipeline.getTrigger().getSubscriptionName());
+    }
+
+    registry.counter(id).increment();
+  }
+
+  private boolean isPubsubTrigger(Trigger trigger) {
+    return PUBSUB_TRIGGER_TYPE.equals(trigger.getType())
+        && (trigger.getSubscriptionName() != null && !trigger.getSubscriptionName().isEmpty())
```
Comment (lwander) [inline echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/PubsubEventMonitor.java]: StringUtils.isEmpty from apache.lang3 I think

File: echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/PubsubEventMonitor.java (hunk 2 above)
Comment (jtk54) [reply echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/PubsubEventMonitor.java]: No, there's one subscription per bucket in the GCS -> Pubsub setup. This is where artifact filters in pipeline triggers come in so you don't trigger on every single change to a bucket. This is obviously not included in this PR -- hopefully arriving shortly.

File: echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/PubsubEventMonitor.java (hunk 3 above)
Comment (jtk54) [reply echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/PubsubEventMonitor.java]: This only applies to a pipeline with a trigger matching `matchTriggerFor` for a pubsub event. This is more or less just an extra guard so we don't try to dereference a field that is null in the Trigger.

File: echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/PubsubEventMonitor.java (hunk 4 above)
Comment (jtk54) [reply echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/PubsubEventMonitor.java]: Right on.

**Pull request merged**